### PR TITLE
C2FTrans: Coarse-to-Fine Transformers for Medical Image Segmentation

### PR DESCRIPTION
**Pull request by HangJung97**

### CODE REVIEW

File: collections/_posts/2022-07-07-C2FTrans-Segmentation.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    F_{ca} &= \text{FFN}(F_{ca}) + F^{'}_3  \\ 
+\end{align}
+$$
+
+To obtain the final CGT output, $$ F_{CGT} $$:
+> $$ F_{ca} \in \mathbb{R}^{\frac{\text{HW}}{16} \times \text{D}} \xrightarrow[\text{reshape}]{} \mathbb{R}^{\text{D} \times \frac{\text{H}}{4} \times \frac{\text{W}}{4}} \xrightarrow[1 \times 1 \text{ conv}]{} F_{CGT} \in \mathbb{R}^{\text{D} \times \frac{\text{H}}{4} \times \frac{\text{W}}{4}} $$
+
+> ⚠️ <span style="color:red">In the paper, it is written that they learn two projection matrices and apply a residual connection to get $$ F_{CGT} $$, which contradicts their GitHub code.</span>
+
+Three functionalities of $$ F_{CGT} $$:
```
Comment (nathanpainchaud) [inline collections/_posts/2022-07-07-C2FTrans-Segmentation.md:67]: From the text before: 

> To obtain the final CGT output, $$ F_{CGT} $$

I understand that F_CGT is the output. But in that case, I'm not sure I get what _functionalities_ an output can have? As in, what _functionalities_ can data have? Would it be more appropriate to talk about what transformations produced the output or downstream tasks? So could you please elaborate on this to make things clearer?

File: collections/_posts/2022-07-07-C2FTrans-Segmentation.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+**3)** Apply non-maximum suppression to keep the highest score windows and discard overlapping boxes. A filtered window set $$ \{w^{*} \} $$ is obtained.  
+**4)** Align $$ \{w^{*} \} $$ with $$ F_2 $$ to form the corresponding feature map $$ \{f^{*} \} $$.
+
+Projections of $$ \{f^{*} \} $$ to obtain $$ Q_{2,k} $$, $$ K_{2,k} $$, and $$ V_{2,k} $$, where $$ k $$ denotes the number of self-attention heads:
+
+$$ \{f^{*} \} \in \mathbb{R}^{\frac{\text{HW}}{4} \times 2\text{C}}  \xrightarrow[\text{tokenization}]{\mathbb{R}^{2 \text{C} \times \text{D}}} \mathbb{R}^{\frac{\text{HW}}{4} \times \text{D}} \xrightarrow[\text{projection}]{\text{E}_{q,k,v} \in \mathbb{R}^{\text(D) \times \text{D}_h}} Q_{2,k}, K_{2,k},  V_{2,k} \in \mathbb{R}^{\frac{\text{HW}}{4} \times \text{D}_h} $$
+
+For each transformer head, the self-attention, $$ F_{sa,k} $$ is computed. Then, the outputs of all head are concatenated to form $$ F_{sa} $$. Just like the CGT, the final BLT output, $$ F_{BLT} $$ will be used to generate the low-res probability map for loss computation and upsampled to produce the full scale probability map.
```
Comment (nathanpainchaud) [inline collections/_posts/2022-07-07-C2FTrans-Segmentation.md:96]: Typo
```suggestion
For each transformer head, the self-attention, $$ F_{sa,k} $$ is computed. Then, the outputs of all the heads are concatenated to form $$ F_{sa} $$. Just like the CGT, the final BLT output, $$ F_{BLT} $$ will be used to generate the low-res probability map for loss computation and upsampled to produce the full scale probability map.
```

File: collections/_posts/2022-07-07-C2FTrans-Segmentation.md [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+   <img src="/collections/images/C2Ftrans/glas.jpg" />
+   <img src="/collections/images/C2Ftrans/acdc_result.jpg" />
+</div>
+
+The authors mention the comparison between the SOTA methods but it lacks the comparison with nnUNet and swin-UNETR, the two most powerful segmentation algorithms.
+
+&nbsp;
+
+# Conclusions
+The introduction of C2F transformer in medical image segmentation is interesting, especially the Cross-scale Global Transformer. However, the paper is not well written as some important details are missing, e.g. the generation and combination of $$ Q $$, $$ K $$, and $$ V $$ of different dimension in the CGT. Moreover, there are some contradictories between some steps described in their paper and GitHub code.
```
Comment (nathanpainchaud) [inline collections/_posts/2022-07-07-C2FTrans-Segmentation.md:134]: Typo + missing words
```suggestion
The introduction of C2F transformer in medical image segmentation is interesting, especially the Cross-scale Global Transformer. However, the paper is not well written as some important details are missing, e.g. the generation and combination of $$ Q $$, $$ K $$, and $$ V $$ of different dimension in the CGT. Moreover, there are some contradictions between some of the steps described in their paper and their GitHub code. 
```

File: collections/_posts/2022-07-07-C2FTrans-Segmentation.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+# Conclusions
+The introduction of C2F transformer in medical image segmentation is interesting, especially the Cross-scale Global Transformer. However, the paper is not well written as some important details are missing, e.g. the generation and combination of $$ Q $$, $$ K $$, and $$ V $$ of different dimension in the CGT. Moreover, there are some contradictories between some steps described in their paper and GitHub code. 
+
+Their GitHub repository is hard to use (incorrect requirements.txt to setup the appropriate virtual environement). I have tested their algorithm on Camus dataset and the results were worse than those given by nnUNet, which made me to doubt the correctness of their results.
+
+
+
+
+
```
Comment (nathanpainchaud) [inline collections/_posts/2022-07-07-C2FTrans-Segmentation.md:141]: Unnecessary blank lines at the end of the file.

**Review by nathanpainchaud — CHANGES_REQUESTED** [review]
Good review. Just a few quick corrections and a slight clarification in one part of the text, and it'll be good to go :slightly_smiling_face:

**Comment by HangJung97** [comment]
Thanks, Nathan. All the suggested changes have been applied! 😀

**Review by nathanpainchaud — APPROVED** [review]
Thansk for the quick turnaround :)

**Pull request merged**